lxd/images.go

```diff
@@ -1115,93 +1115,234 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
-	projectNames := []string{}
+	imageMap := make(map[string][]db.Image)
+
 	err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
-		projects, err := tx.GetProjects(db.ProjectFilter{})
+		var err error
+
+		images, err := tx.GetImages(db.ImageFilter{AutoUpdate: true})
 		if err != nil {
 			return err
 		}
 
-		for _, project := range projects {
-			projectNames = append(projectNames, project.Name)
+		for _, image := range images {
+			imageMap[image.Fingerprint] = append(imageMap[image.Fingerprint], image)
 		}
 
 		return nil
 	})
 	if err != nil {
-		return errors.Wrap(err, "Unable to retrieve project names")
+		return errors.Wrap(err, "Unable to retrieve image fingerprints")
 	}
 
-	for _, project := range projectNames {
-		err := autoUpdateImagesInProject(ctx, d, project)
+	for fingerprint, images := range imageMap {
+		nodes, err := d.cluster.GetNodesWithImageAndAutoUpdate(fingerprint, true)
 		if err != nil {
-			return errors.Wrapf(err, "Unable to update images for project %s", project)
+			logger.Error("Error getting cluster members for image auto-update", log.Ctx{"fingerprint": fingerprint, "err": err})
+			continue
+		}
+
+		if len(nodes) == 0 {
+			continue
+		}
+
+		if len(nodes) > 1 {
+			var nodeIDs []int64
+
+			for _, node := range nodes {
+				err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
+					var err error
+
+					nodeInfo, err := tx.GetNodeByAddress(node)
+					if err != nil {
+						return err
+					}
+
+					nodeIDs = append(nodeIDs, nodeInfo.ID)
+
+					return nil
+				})
+				if err != nil {
+					logger.Error("Unable to retrieve cluster member information for image update", log.Ctx{"err": err})
+					continue
+				}
+			}
+
+			// If multiple nodes have the image, select one to deal with it.
+			if len(nodeIDs) > 1 {
+				selectedNode, err := shared.GetStableRandomInt64FromList(len(images), nodeIDs)
+				if err != nil {
+					logger.Error("Failed to select cluster member for image update", log.Ctx{"err": err})
+					continue
+				}
+
+				// Skip image update if we're not the chosen cluster member.
+				// That way, an image is only updated by a single cluster member.
+				if d.cluster.GetNodeID() != selectedNode {
+					continue
+				}
+			}
+		}
+
+		ch := make(chan struct{})
+		go func() {
+			var deleteIDs []int
+			var newImage *api.Image
+
+			for _, image := range images {
+				_, imageInfo, err := d.cluster.GetImage(image.Project, image.Fingerprint, image.Public)
+				if err != nil {
+					logger.Error("Failed to get image", log.Ctx{"err": err})
+				}
+
+				newInfo, err := autoUpdateImage(d, nil, image.ID, imageInfo, image.Project)
+				if err != nil {
+					logger.Error("Failed to update image", log.Ctx{"err": err})
+				} else {
+					deleteIDs = append(deleteIDs, image.ID)
+				}
+
+				if newImage == nil {
+					newImage = newInfo
+				}
+			}
+
+			if newImage != nil {
+				err := distributeImage(ctx, d, nodes, fingerprint, newImage)
```

Comment:
@monstermunchkin I was more thinking of passing ctx into `autoUpdateImage` and `distributeImage`, and updating the handling here to detect for them returning the `context.Canceled` error and returning at that point rather than using the go routine here. And then inside those two functions adding a comment about needing to handle context properly. That way the plumbing for it is there and this function won't return until those ones have returned cleanly.

That way this function won't have to use ctx.Done() at all.

Reply:
@tomponline Like this? (see last commit)